lib/linked-list/list.rb

```diff
@@ -191,19 +213,39 @@ def to_list
       self
     end
 
-    private
+    def insert_after_node(data, node)
```

Comment:
These three methods `insert_after_node`, `insert_before_node` and `unlink_node` are better to be private I think. Also, let's keep method names similar to others -- `__insert_after`, `__insert_before` and `__unlink`.

Reply:
I need them in conjunction with `#each_node`, as otherwise, I need to go throw list each time, which increase the complexity

Reply:
I'm not sure I understand you meant by this. Would elaborate please?

Reply:
So my use case is to iterate over nodes and by some decision insert data before or after that nodes, like this:
```
list.each_node do |node|
  my_data = ...
  if something?
     list.insert_after_node(node, my_data)
  else
     list.insert_before_node(node, my_data)
  end
end
```

if I do some like this
```
list.each_node do |node|
  my_data = ...
  if something?
     list.insert(my_data, after: ->(d) { node.data == d })
  else
     list.insert(my_data, before: ->(d) { node.data == d })
  end
end
```
it will invoke iteration over list each time and instead of `O(n)` complexity in first case I will get `O(n**2)`

Reply:
Hmm, I only meant to make those methods private and change names to stay consistent with existing naming. That should not affect how you iterate over a list, right?

Reply:
If make them private I should do `list.send(: __insert_after, node, data)` which looks weird, I just need a public API to insert before/after node during `each_node`

Reply:
But where would do that? There used to be something like that in `Node`, but now it is back to `List` which doesn't need explicit caller. Am I missing something?

Reply:
I send you an example of expected from my side usage of linked-list api
```
list.each_node do |node|
  my_data = ...
  if something?
     list.insert_after_node(node, my_data)
  else
     list.insert_before_node(node, my_data)
  end
end
```

Reply:
I see now, so you need both `insert('foo', after: 'bar')` and `insert_after_node(data, node)`. Sure, let's keep them public then, but unlink doesn't have to be. I just don't want to increase public methods surface unnecessarily.

BTW, if it something you are willing to share, what do you use linked list for?

Reply:
I have a schedule data coming from database like: 
```
guy1 busy: [
  { from: '10:30', to: '12:00', reason: 'dinner' }, 
  { from: '9:00', to: '11:00', reason: 'works at object1' }, 
  { from: '0:00', to: '08:00', reason: 'working hours'}, 
  ...
]
``` 

My task is to find all overlapping data points and make them follow one by one so there won't be overlaps, also this data has types which should be merged too

So there are some cases when one data point fully covered by another:
```
slot1: 09:00-13:00, is_home: true, is_in_salon: false 
slot2: 11:00 - 12:00, is_home: true, is_in_salon: true
```

In that case I need to produce 3 slots
```
new_slot1: 09:00-11:00, is_home: true, is_in_salon: false 
new_slot2: 11:00 - 12:00, is_home: true, is_in_salon: true
new_slot3: 12:00 - 13:00, is_home: true, is_in_salon: false
```

as I working with sorted data I do reverse_each(it's my next PR 😉) to find overlaps and do splitting of them and one operation can produce up to 3 new items, to insert 3 elements in the middle of the array I have to resize it and offset tail of array, which is costly operation with linked list it's easy one.

Here is how it looks in the end:

one of my backend test
```
      context do
        let(:working_hours) { { mon: { from: '8:00', to: '20:00' } } }
        let(:home_working_hours) { { mon: { from: '6:00', to: '24:00' } } }
        let!(:schedule_break) do
          create_break_for(
            salon,
            starts_at: date_of_next('mon').change(hour: 4, min: 0),
            ends_at: date_of_next('mon').change(hour: 14, min: 0)
          )
        end

        it do
          get "/vv/owner/salons/#{salon.id}/blocked_times",
              headers: auth_headers(owner),
              params: { date: Date.current }
          expect(blocked_times).to match_array(
            [
              { from: '00:00', to: '04:00', is_home: true, is_in_salon: true, reasons: ['Not working hours'] },
              { from: '04:00', to: '06:00', is_home: true, is_in_salon: true, reasons: ['Not working hours', "Schedule break '#{schedule_break.name}'"] },
              { from: '06:00', to: '08:00', is_home: true, is_in_salon: true, reasons: ['Not working hours', "Schedule break '#{schedule_break.name}'"] },
              { from: '08:00', to: '14:00', is_home: true, is_in_salon: true, reasons: ["Schedule break '#{schedule_break.name}'"] },
              { from: '20:00', to: '23:59', is_home: false, is_in_salon: true, reasons: ['Not working hours'] }
            ]
          )
        end
      end
```

and on front-end, I need to display this data like this:
 
![image](https://user-images.githubusercontent.com/1548210/44745500-91475e00-ab10-11e8-98e8-306847bc482d.png)

Reply:
That's very cool, man! Thanks for sharing!